src/main/kotlin/com/team4099/robot2023/subsystems/vision/Vision.kt

```diff
@@ -1,18 +1,286 @@
+import java.util.function.Supplier
 
-class Vision(cameras: VisionIO) : SubsystemBase() {
-  val ambiguityThreshold = 0.15
-  val targetLogTime = 0.05.seconds
-  val cameraPoses = VisionConstants.CAMERA_TRANSFORMS
+class Vision(vararg cameras: CameraIO) : SubsystemBase() {
+  val io: List<CameraIO> = cameras.toList()
+  val inputs = List(io.size) { CameraIO.CameraInputs() }
 
+  companion object {
+    val ambiguityThreshold = 0.15
+    val targetLogTime = 0.05.seconds
+    val cameraPoses = VisionConstants.CAMERA_TRANSFORMS
+
+    val xyStdDevModel =
+      PolynomialRegression(
+        doubleArrayOf(
+          0.752358,
+          1.016358,
+          1.296358,
+          1.574358,
+          1.913358,
+          2.184358,
+          2.493358,
+          2.758358,
+          3.223358,
+          4.093358,
+          4.726358
+        ),
+        doubleArrayOf(
+          0.005, 0.0135, 0.016, 0.038, 0.0515, 0.0925, 0.0695, 0.046, 0.1245, 0.0815, 0.193
+        ),
+        1
+      )
+
+    val thetaStdDevModel =
+      PolynomialRegression(
+        doubleArrayOf(
+          0.752358,
+          1.016358,
+          1.296358,
+          1.574358,
+          1.913358,
+          2.184358,
+          2.493358,
+          2.758358,
+          3.223358,
+          4.093358,
+          4.726358
+        ),
+        doubleArrayOf(
+          0.008, 0.027, 0.015, 0.044, 0.04, 0.078, 0.049, 0.027, 0.059, 0.029, 0.068
+        ),
+        1
+      )
+  }
+
+  private var poseSupplier = Supplier<Pose2d> { Pose2d() }
+  private var visionConsumer: Consumer<List<PoseEstimator.TimestampedVisionUpdate>> = Consumer {}
+  private val lastFrameTimes = mutableListOf<Time>()
+  private val lastTagDetectionTimes = mutableMapOf<Int, Time>()
+
+  init {
+    for (i in 0..io.size) {
+      lastFrameTimes[i] = 0.0.seconds
+    }
+
+    FieldConstants.aprilTags.map { lastTagDetectionTimes[it.id] = 0.0.seconds }
+  }
+
+  fun setDataInterfaces(
+    poseSupplier: Supplier<Pose2d>,
+    visionConsumer: Consumer<List<PoseEstimator.TimestampedVisionUpdate>>
+  ) {
+    this.poseSupplier = poseSupplier
+    this.visionConsumer = visionConsumer
+  }
+
+  override fun periodic() {
+    for (i in 0..io.size) {
+      io[i].updateInputs(inputs[i])
+      Logger.getInstance().processInputs("Vision/${VisionConstants.CAMERA_NAMES[i]}", inputs[i])
+    }
+
+    val currentPose: Pose2d = poseSupplier.get()
+    val robotPoses = mutableListOf<Pose2d>()
+    val visionUpdates = mutableListOf<PoseEstimator.TimestampedVisionUpdate>()
+
+    for (instance in 0..io.size) {
+
+      for (frameIndex in 0..inputs[instance].timestamps.size) {
+        lastFrameTimes[instance] = Clock.fpgaTime
+        val timestamp = inputs[instance].timestamps[frameIndex]
+        val values = inputs[instance].frames[frameIndex]
+
+        var cameraPose: Pose3d? = null
+        var robotPose: Pose2d? = null
+
+        when (values[0]) {
+          1.0 -> {
+            cameraPose =
```

Comment:
More broadly, I think it would be better if the cameras (the IO classes) spit out a list of `TimestampedVisionPose`s (or whatever they're called) directly, rather than giving you a list of numbers to interpret. This is like the perfect example of implementation specific stuff -- the limelight won't give you the same array

Reply:
The ambiguity check etc can be done in here, I think, but it would be very nice if the IO objects passed over data as actual poses

Reply:
In other words: pass the same data we are passing, just encode it into a usable format (poses/transforms/whatever)